src/auth/github-auth.service.ts

```diff
@@ -0,0 +1,99 @@
+import { GithubUser } from "../global/types/auth.types";
+import { generatePasswordHash } from "../global/auth/utils";
+import { OAuthConfig } from "@/config/Oauth/oauthConfig.interface";
+
+@Injectable()
+export class GithubAuthService implements IAuthProvider {
+    constructor(
+        private prisma: PrismaService,
+        @Inject("OAuth-Config") private oAuthConfig: OAuthConfig,
+    ) {}
+    async validateUser(user: GithubUser) {
+        const userInDb = await this.prisma.findUserByOAuthId(
+            "github",
+            user.githubId,
+        );
+
+        if (userInDb)
+            return {
+                id: userInDb.userId,
+                email: user.email,
+            };
+        return this.createUser(user);
+    }
+
+    async createUser(user: GithubUser): Promise<AuthUserResult> {
+        // generate a random password and not tell them so they can't login, but they will be able to reset password,
+        // and will be able to login with this in future,
+        // or maybe the app will prompt user the input the password, exact oauth flow is to be determined
+
+        // this should not happen when "user:email" is in the scope
+        if (!user.email)
+            throw new InternalServerErrorException(
+                "[github-auth.service]: Cannot get email from github to create a new Chingu account",
+            );
+
+        // check if email is in the database, add oauth profile to existing account, otherwise, create a new user account
+        let upsertResult;
+        try {
+            upsertResult = await this.prisma.user.upsert({
+                where: {
+                    email: user.email,
+                },
+                update: {
+                    emailVerified: true,
+                    oAuthProfiles: {
+                        create: {
+                            provider: {
+                                connect: {
+                                    name: "github",
+                                },
+                            },
+                            providerUserId: user.githubId,
+                            providerUsername: user.username,
+                        },
+                    },
+                },
+                create: {
+                    email: user.email,
+                    password: await generatePasswordHash(),
+                    emailVerified: true,
+                    avatar: user.avatar,
+                    oAuthProfiles: {
+                        create: {
+                            provider: {
+                                connect: {
+                                    name: "github",
+                                },
+                            },
+                            providerUserId: user.githubId,
+                            providerUsername: user.username,
+                        },
+                    },
+                },
+            });
+        } catch (e) {
+            if (e.code === "P2025") {
```

Comment:
What happens if the code is not P2025 (PrismaClientKnownRequestError)? I don't see an error being thrown for any other kind of errors in this catch statement

Reply:
if it's not a P2025 error it would just re-throw the error which is handled in the `prisma-client-exception.filter.ts` Exception filter if it's a prisma exception. if not it would just be catched by the default nest.js exception filter https://docs.nestjs.com/exception-filters